ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,312 @@
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            outFile << "EQUILIBRIUM_PHASES " << node_id + 1 << "\n";
+            outFile << nodal_equilibrium_phases << "\n";
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            outFile << "KINETICS " << node_id + 1 << "\n";
+            outFile << nodal_kinetic_reactant;
+            outFile << "-steps " << phreeqcIO._dt << "\n" << "\n";
+        }
+
+        if (node_id == 0)
+        {
+            auto const& reaction_rates = phreeqcIO._reaction_rates;
+            if (!reaction_rates.empty())
+            {
+                outFile << "RATES" << "\n";
+                outFile << reaction_rates << "\n";
+            }
+
+            outFile << "SELECTED_OUTPUT" << "\n";
+            outFile << *phreeqcIO._output;
+        }
+
+        outFile << "END" << "\n" << "\n";
+    }
+
+    return outFile;
+}
+
+void phreeqcIO::execute()
+{
+    auto database_loaded = [&](int instance_id) {
+        if (LoadDatabase(instance_id, _database.c_str()) == 0)
+            return true;
+        else
+            return false;
+    };
+
+    DBUG("Phreeqc: Executing chemical calculation.");
+    // initialize phreeqc configurations
+    auto const instance_id = CreateIPhreeqc();
+
+    // load a specific database in the working directory
+    if (!database_loaded(instance_id))
+        ERR("Failed to load the specified thermodynamic database file %s.",
+            _database.c_str());
+
+    SetSelectedOutputFileOn(instance_id, 1);
+
+    if (RunFile(instance_id, _input_file) > 0)
+        OutputErrorString(instance_id);
+}
+
+void phreeqcIO::readOutputsFromFile()
+{
+    auto const& basic_output_setups = _output->basic_output_setups;
+    auto const& output_file = basic_output_setups.getOutputFile();
+    DBUG("Reading phreeqc outputs from file '%s'.", output_file);
+    std::ifstream inFile(output_file, std::ifstream::in);
+
+    if (!inFile.good())
+        ERR("Could not open file %s.", output_file);
+
+    inFile >> *this;
+    inFile.close();
+}
+
+std::ifstream& operator>>(std::ifstream& inFile, phreeqcIO& phreeqcIO)
+{
+    // Skip first two lines, i.e., headline and equilibrium calculation result
+    // of initial solution whose id is 1.
+    for (int i = 0; i < 2; ++i)
+        inFile.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+
+    std::string line;
+    boost::char_separator<char> sep(" ");
+    auto const& output = *phreeqcIO._output;
+    auto const& dropped_item_ids = output.dropped_item_ids;
+    int node_id = 0;
+    while (std::getline(inFile, line))
+    {
+        std::vector<double> accepted_items;
+        boost::tokenizer<boost::char_separator<char>> items(line, sep);
+        for (boost::tokenizer<boost::char_separator<char>>::iterator item_iter =
```

Comment:
Isn't the boost tokenizer supporting a for-range loop?

Reply:
Pleas use `boost::algorithm::split`, with which you can even use for-range loop.

Reply:
Resolved.